x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/DelegatingProcessor.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.common;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Processor that delegates the {@link java.util.concurrent.Flow.Subscription} to the upstream {@link java.util.concurrent.Flow.Publisher}
+ * and delegates event transmission to the downstream {@link java.util.concurrent.Flow.Subscriber}.
+ */
+public abstract class DelegatingProcessor<T, R> implements Flow.Processor<T, R> {
+    private static final Logger log = LogManager.getLogger(DelegatingProcessor.class);
+    private final AtomicLong pendingRequests = new AtomicLong();
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private Flow.Subscriber<? super R> downstream;
+    private Flow.Subscription upstream;
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super R> subscriber) {
+        if (downstream != null) {
+            subscriber.onError(new IllegalStateException("Another subscriber is already subscribed."));
+            return;
+        }
+
+        downstream = subscriber;
+        downstream.onSubscribe(forwardingSubscription());
+    }
+
+    private Flow.Subscription forwardingSubscription() {
+        return new Flow.Subscription() {
+            @Override
+            public void request(long n) {
+                if (isClosed.get()) {
+                    downstream.onComplete(); // shouldn't happen, but reinforce that we're no longer listening
+                } else if (upstream != null) {
+                    upstream.request(n);
+                } else {
+                    pendingRequests.accumulateAndGet(n, Long::sum);
+                }
+            }
+
+            @Override
+            public void cancel() {
+                if (isClosed.compareAndSet(false, true) && upstream != null) {
+                    upstream.cancel();
+                }
+            }
+        };
+    }
+
+    @Override
+    public void onSubscribe(Flow.Subscription subscription) {
+        if (isClosed.get()) {
```

Comment:
Sure we can - something to say this subscriber only supports events from a single publisher (because it only supports publishing to a single subscriber).

If we wanted many-to-many that would likely be another PR/object